apps/confessions-server/src/routing/routes/confessionRoutes.ts

```diff
@@ -0,0 +1,152 @@
+import { ApplicationContext } from "../../types";
+import { SEMAPHORE_GROUP_URL, ACCESS_TOKEN_SECRET, authenticateJWT } from "../../util/auth";
+import { prisma } from "../../util/prisma";
+
+export function initConfessionRoutes(
+  app: express.Application,
+  _context: ApplicationContext
+): void {
+  // this endpoint verifies the proof is valid,
+  // and the proof claim matches the semaphoreGroupUrl and the confession in the request
+  // before saving the confession to the database.
+  app.post("/confessions", async (req: Request, res: Response, next: NextFunction) => {
+    const request = req.body as PostConfessionRequest;
+
+    try {
+     const err = await verifyGroupProof(request.semaphoreGroupUrl, request.proof, request.confession)
+     if (err != null) throw err;
+
+      // proof should be unique
+      await prisma.confession.upsert({
+        where: {
+          proof: request.proof
+        },
+        update: {},
+        create: {
+          semaphoreGroupUrl: request.semaphoreGroupUrl,
+          body: request.confession,
+          proof: request.proof
+        }
+      });
+
+      res.send("ok");
+    } catch (e) {
+      console.error(e);
+      next(e);
+    }
+  });
+
+  // this login endpoint and the get confessions endpoint below shows how an auth flow could work.
+  // this endpoint verfies the proof and returns a jwt if succeed.
+  // the client can then use this jwt in its future requests.
+  // the get confessions endpoint uses the middleware authenticateJWT to authenticate jwt.
+  app.post("/login", async (req: Request, res: Response, next: NextFunction) => {
+    const request = req.body as LoginRequest;
+
+    try {
+      const err = await verifyGroupProof(request.semaphoreGroupUrl, request.proof)
+      if (err != null) throw err;
+
+      const accessToken = sign({ groupUrl: request.semaphoreGroupUrl }, ACCESS_TOKEN_SECRET)
+
+      res.status(200).json({ accessToken });
+    } catch (e) {
+      console.error(e);
+      next(e);
+    }
+  });
+
+  app.get("/confessions", authenticateJWT, async (req: Request, res: Response) => {
+    const page = queryStrToInt(
+      req.query.page,
+      1,
+      (i: number) : boolean => { return i >= 1}
+    );
+    const limit = queryStrToInt(
+      req.query.limit,
+      20,
+      (i: number) : boolean => { return i >= 1}
+    );
+
+    const confessions = await prisma.confession.findMany({
+      take: limit,
+      skip: (page - 1) * limit,
+      orderBy: { updatedAt: "desc" },
+    });
+    res.status(200).json({ confessions });
+  });
+}
+
+export interface LoginRequest {
+  semaphoreGroupUrl: string;
+  proof: string;
+}
+
+export interface PostConfessionRequest {
+  semaphoreGroupUrl: string;
+  confession: string;
+  proof: string;
+}
+
+async function verifyGroupProof(
+  semaphoreGroupUrl: string,
+  proof: string,
+  signal?: string
+): Promise<Error | null> {
+  // only allow Zuzalu group for now
+  if (semaphoreGroupUrl !== SEMAPHORE_GROUP_URL) {
```

Comment:
maybe we could make `SEMAPHORE_GROUP_URL` load from an environment variable, and document that environment variable. this would make it much easier for a newbie dev to clone this repo and make it work with some other arbitrary semaphore group.

Reply:
made SEMAPHORE_GROUP_URL to load from env var, also added the documentation